src/main/java/org/graylog/plugins/netflow/codecs/NetflowV9CodecAggregator.java

```diff
@@ -0,0 +1,217 @@
+                        // the loader cannot fail, it only creates a new queue
+                    }
+                } else {
+                    // include the template in our result
+                    templates.add(templateKey);
+                    packetsToSend.add(buf);
+                }
+            });
+
+            // if we have buffered this packet, don't try to process it now. we still need all the templates for it
+            if (packetBuffered[0]) {
+                return new Result(null, true);
+            }
+
+            // if we didn't buffer anything but also didn't have anything queued that can be processed, don't proceed.
+            if (packetsToSend.isEmpty()) {
+                return new Result(null, true);
+            }
+
+            // if we have any packets to forward, prepare a result
+            final ChannelBuffer resultBuffer = ChannelBuffers.dynamicBuffer();
+            resultBuffer.writeByte(NetFlowCodec.ORDERED_V9_MARKER);
+
+            final NetFlowV9Journal.RawNetflowV9.Builder builder = NetFlowV9Journal.RawNetflowV9.newBuilder();
+
+            // add the used templates and option template to the journal message builder
+            templates.stream()
+                    .map(templateKey -> Maps.immutableEntry(templateKey, templateCache.getIfPresent(templateKey)))
+                    .forEach(entry -> {
+                        final TemplateKey templateKey = entry.getKey();
+                        final TemplateByteBuf templateByteBuf = entry.getValue();
+                        if (templateByteBuf == null) {
+                            LOG.warn("Template {} expired while processing, discarding netflow packet", templateKey);
+                        } else if (templateByteBuf.optionTemplate) {
+                            LOG.debug("Writing options template flow {}", templateKey);
+                            builder.putOptionTemplate(1, ByteString.copyFrom(templateByteBuf.buf.nioBuffer()));
+                        } else {
+                            LOG.debug("Writing template {}", templateKey);
+                            builder.putTemplates(templateKey.getTemplateId(), ByteString.copyFrom(templateByteBuf.buf.nioBuffer()));
+                        }
+                    });
+
+            // finally write out all the packets we had buffered as well as the current one
+            packetsToSend.forEach(packetBuffer -> builder.addPackets(ByteString.copyFrom(packetBuffer.toByteBuffer())));
+            resultBuffer.writeBytes(builder.build().toByteArray());
+            return new Result(resultBuffer, true);
+
+        } catch (Exception e) {
+            LOG.error("Unexpected failure while aggregating NetFlowV9 packet, discarding packet.", ExceptionUtils.getRootCause(e));
+            return new Result(null, false);
+        }
+    }
+
+    private void queueBufferedPackets(Set<TemplateKey> templates, Set<ChannelBuffer> packetsToSend, TemplateKey templateKey) {
```

Comment:
This actually requires checking all templates used by a packet, so this code will change soon and is not the final version.